Add new blog: Kvrocks 2023 In Review (#187)
apache/kvrocks-website: blog/2024-01-07-kvrocks-2023-in-review/index.md

---
slug: kvrocks-2023-in-review
title: Apache Kvrocks 2023 In Review
---

The year 2023 is a small milestone for the community in open source, and it is also a new starting point. **Kvrocks successfully graduated from the incubator to become an Apache top-level project in June, which is a great affirmation for the community's health and sustainability**. At the same time, there are also some exciting progress in the project and community.

> Beofre that, we would like to thanks all the contributors and users who have been supporting Apache Kvrocks. We can't make it without you!

Reviewer: Before?

> Beofre we start, we would like to thanks all the contributors and users who have been supporting Apache Kvrocks. We can't make it without you!

Reviewer 2: ```suggestion
Before we start, we would like to thanks all the contributors and users who have been supporting Apache Kvrocks. We can't make it without you!
```

<!--truncate-->

### Project Retrospective

> We collected a lot of feedback from users and developers in the discussion: [2023 Planning(Want your feedback)](https://github.com/apache/kvrocks/discussions/1226), some of them have been resolved in 2023, and some of them are still in the plan list:

Reviewer 2: ```suggestion
We collected a lot of feedback from users and developers in the discussion: [2023 Planning(Want your feedback)](https://github.com/apache/kvrocks/discussions/1226), some of them have been resolved in 2023, and some of them are still in the plan list.
```

**The key progresses in 2023:**

- JSON data structure
- BloomFilter data structure
- Watch/Unwatch command
- Lua Functions
- Replication with TLS

> - 65 bits expire time and size

Reviewer: Can we describe it more specificly?

Author: Thank you!

> - 65 bits expire time and size

Author: ```suggestion
- Allow to enable 64 bits expiration time(precision is millisecond instead of second), and the length of List/Hash/Set/ZSET/Stream can be also extended from 32 to 64 bits.
```
@mapleFU Is this more clear than before?

Reviewer: Yeah I think so

- Allow to use `LOAD RDB` command to load Redis RDB file
- Allow to dynamically adjust the number of Worker threads
- Enable Namespace replication
- Support RESTORE command

**What we are still working on:**

- RESP3 protocol, more information can be found in [Tracking issue: Support RESP3 in Kvrocks](https://github.com/apache/kvrocks/issues/1980)
- Kvrocks in Kubernetes, [kvrocks-operator](https://github.com/RocksLabs/kvrocks-operator) is ready to test now, but still need more work to make it production-ready
- Controller to manage cluster, the API part of [kvrocks-controller](https://github.com/RocksLabs/kvrocks-controller) is finished, but we belive it's not enough of users before UI part is ready

**What still in the backlog:**

- Allow to use the Raft protocol for replication

> - HyperLogLog/RedisGraph data structure

Reviewer 2: RedisGraph itself is already not in active development, so I'm not sure if we need to put it in an important place.

Reviewer 2: Instead, from my perspective, RedisSearch is more important and has a widely user group, especially considering the vector search. Also RedisTimeSeries is good to me.

> - HyperLogLog/RedisGraph data structure

Author: ```suggestion
- HyperLogLog
- RedisGraph, no plan to support since it's deprecated by the Redis community.
```

Author: @PragmaTwice How do you think about this description?

- Semi-sync replication

> - Client-side caching

Reviewer 2: IMHO it is hard for our to maintain clients instead of using these Redis clients directly.

Author: Client-side caching should only be compatible with the RESP3 "PUSH" protocol which can notify the client side if the key was modified. But supporting the RESP3 "PUSH" is also NOT easy work for Kvrocks.

Reviewer 2: Ahhh I see. Thank you.

**The key progresses in 2023:**

- JSON data structure
- BloomFilter data structure
- Watch/Unwatch command
- Lua Functions
- Replication with TLS
- Allow to enable 64 bit expiration time (precision is in milliseconds instead of seconds), and the length of List/Hash/Set/ZSET/Stream can be also extended from 32 to 64 bits.
- Allow to use `LOAD RDB` command to load Redis RDB file
- Allow to dynamically adjust the number of Worker threads
- Enable Namespace replication
- Support RESTORE command

**What we are still working on:**

- RESP3 protocol, more information can be found in [Tracking issue: Support RESP3 in Kvrocks](https://github.com/apache/kvrocks/issues/1980)
- Kvrocks in Kubernetes, [kvrocks-operator](https://github.com/RocksLabs/kvrocks-operator) is ready to test now, but still need more work to make it production-ready
- Controller to manage cluster, the API part of [kvrocks-controller](https://github.com/RocksLabs/kvrocks-controller) is finished, but we belive it's not enough of users before UI part is ready

**What still in the backlog:**

- Allow to use the Raft protocol for replication
- HyperLogLog

> - RedisGraph, no plan to support since it's deprecated by the Redis community.

Author: ```suggestion
- RedisGraph, no plan to support since it's deprecated by the Redis community
```

- Watch/Unwatch command
- Lua Functions
- Replication with TLS

> - Allow to enable 64 bits expiration time(precision is millisecond instead of second), and the length of List/Hash/Set/ZSET/Stream can be also extended from 32 to 64 bits.

Reviewer 2: ```suggestion
- Allow to enable 64 bit expiration time (precision is in milliseconds instead of seconds), and the length of List/Hash/Set/ZSET/Stream can be also extended from 32 to 64 bits.
```
